plugins/basic/nexus-repository-docker/src/main/java/org/sonatype/nexus/repository/docker/internal/DockerFacetImpl.java

```diff
@@ -0,0 +1,682 @@
+
+    // set image name and tag
+    tagComponent.formatAttributes().set(P_IMAGE_NAME, name);
+    tagComponent.formatAttributes().set(P_IMAGE_TAG, tag);
+    // set layer reference
+    tagComponent.formatAttributes().set(P_LAYER_ID, layerId);
+    // copy ancestry from referenced layer (so we can search images that include the layer)
+    tagComponent.formatAttributes().set(P_LAYER_ANCESTRY, metadataAsset.formatAttributes().get(P_ANCESTRY));
+    // set docker daemon metadata
+    NestedAttributesMap tagDaemonMetadata = tagComponent.formatAttributes().child(P_DAEMON_METADATA);
+    tagDaemonMetadata.clear();
+    for (Entry<String, String> entry : daemonMetadata.entrySet()) {
+      tagDaemonMetadata.set(entry.getKey(), entry.getValue());
+    }
+    tx.saveComponent(tagComponent);
+  }
+
+  /**
+   * Find a component by its name and tag (version)
+   *
+   * @param tx   storage transaction
+   * @param name image name
+   * @param tag  image tag
+   * @return found component of null if not found
+   */
+  @Nullable
+  private Component findComponent(final StorageTx tx, final String name, final String tag) {
+    ComponentQuery query = new Builder()
+        .where(P_NAME + "=").param(name)
+        .where(" AND " + P_VERSION + "=").param(tag)
+        .build();
+    Iterable<Component> components = tx.findComponents(
+        query.getWhere(), query.getParameters(), singletonList(getRepository()), null
+    );
+    if (components.iterator().hasNext()) {
+      return components.iterator().next();
+    }
+    return null;
+  }
+
+  /**
+   * Find all components (tags) given a component name.
+   *
+   * @param tx   storage transaction
+   * @param name image name
+   * @return found components
+   */
+  private Iterable<Component> findComponents(final StorageTx tx, final String name) {
+    ComponentQuery query = new Builder()
+        .where(P_NAME + "=").param(name)
+        .build();
+    return tx.findComponents(
+        query.getWhere(), query.getParameters(), singletonList(getRepository()), null
+    );
+  }
+
+  /**
+   * Find an asset by its name.
+   *
+   * @param tx        storage transaction
+   * @param bucket    repository bucket
+   * @param assetName asset name
+   * @return found asset or null if not found
+   */
+  @Nullable
+  private Asset findAsset(final StorageTx tx, final Bucket bucket, final String assetName) {
+    return tx.findAssetWithProperty(StorageFacet.P_NAME, assetName, bucket);
+  }
+
+  /**
+   * Validate that layer metadata is a proper JSON map.
+   *
+   * @param layerMetadataContent layer metadata content
+   * @return parsed layer metadata
+   */
+  private Map<String, Object> readLayerMetadata(final Supplier<InputStream> layerMetadataContent) {
+    TypeReference<Map<String, Object>> typeRef = new TypeReference<Map<String, Object>>() {}; //CHECKSTYLE:OFF
+    try {
+      return mapper.readValue(layerMetadataContent.get(), typeRef);
+    }
+    catch (Exception e) {
+      log.debug("Invalid JSON", e);
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Define a constant instead of duplicating this literal "Invalid JSON" 3 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1192)